### PR TITLE
Add an exception message mapping and write exception message to the error file

### PR DESCRIPTION
**Pull request by ohchocolate**
### Description of changes
* Modify the message_troubleshooting_link to provide a generic url for users to reference.
* Add a mapping that catches "Cluster has been set to PROTECTED mode due to failures detected in static node provisioning" exception and consequently displays a specific error message associated with the CloudFormation stack.

### Tests
* Create clusters that produce the protected mode error with custom cookbook and see the error message in the CloudFormation stack 
Screenshot 2023-06-01 at 1 27 41 PM

* SSH into the head node and see the added error message in the bootstrap_error_msg

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb [hunk 1]
```diff
@@ -47,14 +47,21 @@ def report
...
+
         failed_action_collection.each do |action_record|
           # there might be multiple failed action records
           # here we only look at the outermost layer resource by setting nesting_level = 0
           # with the assumption that there is only one failed action record with nesting_level = 0
           next unless action_record.nesting_level == 0
           # we first check if it is a storage mounting error for EBS, RAID, EFS, or FSX,
           # otherwise we will get the recipe information
-          message_error = mount_message_mapping[action_record.new_resource.name] || "Failed to run chef recipe#{get_recipe_info(action_record)}."
+          message_error =exception_message_mapping[action_record.exception.message] ||
```
Comment (chenwany) [inline cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: missing a space between "=" and "e"

File: cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb [hunk 2]
```diff
@@ -47,14 +47,21 @@ def report
           "mount fsx" => "Failed to mount FSX.",
         }
 
+        # define a mapping from the exception information to the error message we would like to display
+        exception_message_mapping = {
+          "Cluster has been set to PROTECTED mode due to failures detected in static node provisioning" =>
+            "Cluster has been set to PROTECTED mode due to failures detected in static node provisioning."
```
Comment (chenwany) [inline cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: Missing "," at the end

**Review by chenwany — COMMENTED** [review]
Could you fix the formatting issues mentioned in the ruby code check?

**Review by delongmeng — DISMISSED** [review]
Some Ruby offenses need to be fixed. (see the `Some checks were not successful` section below)

File: cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb (hunk 2 above)
Comment (delongmeng) [reply cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: Should it be a period (".") or comma at the end? (never mind if the original exception message does not include the ending period)

BTW, maybe we can define a variable for the protected mode first to make it slightly more readable? something like this:
```
protected_mode_exception = "Cluster has been set to PROTECTED mode due to failures detected in static node provisioning"
exception_message_mapping = {
  protected_mode_exception => protected_mode_exception + ".",
}
```

what do you think @chenwany ?

Comment (delongmeng) [reply cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: Oh I was referring to the end of  "node provisioning" in the key part, and maybe you were referring to the missing comma after the whole key-value pair (after "node provisioning.")?

Comment (delongmeng) [reply cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: Can you also update the screenshot to include the ending period of this error message sentence @ohchocolate ?

**Review by delongmeng — COMMENTED** [review]
Minor: For the description part, maybe you can combine the 1st and 3rd points? Something like (just an example):
```
Add a mapping to detect the PROTECTED mode exception and display the specific error message accordingly to CloudFormation stack.
```

File: cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb (hunk 2 above)
Comment (chenwany) [reply cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: Yeah my comment is about adding `,` after `."`, about the formatting issue
Yes I think Delong's comment make sense too

File: cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb (hunk 1 above)
Comment (ohchocolate) [reply cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: Thanks! Done

File: cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb (hunk 2 above)
Comment (ohchocolate) [reply cookbooks/aws-parallelcluster-config/libraries/write_chef_error_handler.rb]: I revised the mapping as Delong suggested and fixed the format. Thanks!

**Review by chenwany — APPROVED** [review]
lgtm

**Comment by ohchocolate** [comment]
> Some Ruby offenses need to be fixed. (see the `Some checks were not successful` section below)

Fixed the Ruby style problem.

**Pull request merged**